Spin SimModel off NeurodamusModel and build standalone nrnmech lib

ferdonline opened the pull request:
#### Problem
Neurodamus has traditionally been built together with simulation models. However, if a user wanted to build the model alone he would have to filter the mod files of interest and compile himself.
The first part of the problem was addressed before with the new neurodamus split repositories. However, until now, the spack neurodamus recipes merge all the files again to build the final complete neurodamus `special`. Nevertheless some users require the standalone model `libnrnmech` for their own custom simulation programs. In these cases they would have still to compile the mods themselves.

#### Introducing standalone simulation model packages
Standalone simulation models are packages of mechanisms for neuron, which the user can apply directly in their own simulation programs without neurodamus. NeurodamusModel was refactored to build on SimModel, deploying standalone libnrnmech and libnrnmech_damus featuring required neurodamus mods.

**Output:**
- lib/libnrnmech.so: The original mechanism library (exported as env var LIBNRNMECH_PATH)
- lib/libnrnmech_nd.so: The neurodamus library with mechanisms
- lib/special: The neurodamus launcher, using libnrnmech_nd.so

Besides being exported as `LIBNRNMECH_PATH`, the mech lib path is also exported as `BGLIBPY_MOD_LIBRARY_PATH` for temporary compatibility.

ferdonline committed:
Introducing standalone simulation models

Standalone simulation models are packages of mechanisms for neuron
Being standalone, the user can use them directly in their own
simulation programs wihtout neurodamus, and they set the foundation
for bglibpy and future decoupled neurodamus.

In this initial version 'neocortex' is being added

ferdonline commented on var/spack/repos/builtin/packages/sim-model/package.py:
```diff
@@ -0,0 +1,74 @@
+            shutil.move(cmod, prefix.share.modc)
+
+    def setup_environment(self, spack_env, run_env):
+        run_env.set('BGLIBPY_MOD_LIBRARY_PATH',
```
As discussed, it would be nice that at some point this env var is renamed to not be tied to BglibPy

pramodk reviewed (dismissed):
we discussed this off-line : we will refactor existing neurodamus modules to allow this flexibility.

ferdonline committed:
Nurodamus-model to install bare sim models

In this refactoring we take into consideration that it might tricky to
maintain 2* models (neurodamus/bare), and that having a single neurodamus that
dlopens the model might result in a fragile setup.

In this approach, neurodamus-model extends sim-model and takes the
required steps to install a bare sim model (libnrnmech / libcorenrnmech)
but proceeds to install the combined neurodamus+model with the advantage
of doing an incremental compilation (all mod2c output and .o files are
reused) both for neuron and coreneuron builds.
The resulting libs (libnrnmech_damus / libcorenrnmech_*_damus) are deployed
besides the original ones, while the new special/special-core replace
the original "bare" specials

Witht this reorganization, the logic was largely broken down and split
between mod compilation (in sim-model) and combining the model with neurodamus
sources and dependencies.

ferdonline committed:
Dropping bare model-neocortex

ferdonline committed:
Removing old neurodamus from deployments

ferdonline committed:
Merge remote-tracking branch 'origin/develop' into new/standalone_models

ferdonline committed:
libnrnmech path exported as LIBNRNMECH_PATH

We also provide BGLIBPY_MOD_LIBRARY_PATH for compat but it's deprecated.

ferdonline commented:
Dont merge yet, not passing SimulationStack tests...

ferdonline committed:
special bash-version: preserve arguments to nrniv

ferdonline committed:
Again LC_ALL. Fixed in neuron but not deployed yet

ferdonline commented:
Now passing SimulationStack:
https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/499/pipeline

Bug was affecting all '~binary' installations.

pramodk commented on deploy/packages/bbp-packages.yaml:
```diff
@@ -116,11 +116,6 @@ packages:
-      - neurodamus+coreneuron^coreneuron+knl
       - neurodamus-neocortex
       - neurodamus-hippocampus
       - neurodamus-thalamus
```
additional `+knl`  are needed

ferdonline replied:
see the whole file. There are +coreneuron+knl versions
Or is still any combination missing? I didn't delete anything else...

pramodk replied:
Ahh, now I see that! all good!

pramodk commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -99,69 +79,31 @@ def build(self, spec, prefix):
-                prefix.bin.special)
+        # base dest dirs already created by model install
+        arch = arch = spec.architecture.target
+        dst_libname = 'libnrnmech_damus.so'
```
naming question again :D  why  not full `neurodamus` instead of `damus` ?  I don't think there is any meaning to `damus` :)

ferdonline replied:
it's a shorthand. Is it too bad? Sure we can change
Edit: I made it libnrnmechdamus. I don't think there's the need to have a long/fancier name, for developers it's clear enough

wvangeit replied:
Does this actually need to be renamed ?

ferdonline replied:
This is the library internal to neurodamus (which includes aux mods). Has to be renamed to avoid overwriting the original libnrnmech which is what you need, right?

wvangeit replied:
if there is an libnrnmech in the same dir, then yes, renaming doesn't make sense.
Otherwise, I'm not sure if it couldn't have the same name.

ferdonline replied:
I guess there's some confusion here. The recipe is renaming the new one.
The env var `BGLIBPY_MOD_LIBRARY_PATH` will point to the original `libnrnmech.so` which doesnt contain the the util mods

ferdonline replied:
I updated the PR description to be clearer in this regard.

pramodk commented on var/spack/repos/builtin/packages/sim-model/package.py:
```diff
@@ -0,0 +1,147 @@
+    phases = ('build', 'install')
+
+    # The name of the mechanism, which cen be overriden
+    mech_name = ''
```
what is this for?

ferdonline replied:
subclasses override, and superclass uses it for lib name

pramodk commented on var/spack/repos/builtin/packages/sim-model/package.py:
```diff
@@ -0,0 +1,147 @@
+            mechlib = self.__build_mods_coreneuron(
+                corenrn_mods or mods_location, link_flag, include_flag)
+            # Link neuron special with this mechs lib
+            link_flag += ' ' + mechlib.ld_flags + self._get_lib_flags('coreneuron')
```
better to put empty space between `mechlib.ld_flags + self._get_lib_flags` ?

ferdonline replied:
self._get_lib_flags already outputs with req. spaces
Edit: I changed to be more explicit as you said

pramodk commented on var/spack/repos/builtin/packages/sim-model/package.py:
```diff
@@ -0,0 +1,147 @@
+
+    conflicts('^neuron~python', when='+coreneuron')
+
+    # -
```
curious : `# -` means?

ferdonline replied:
Just to add more space in a pep8 valid way. yeah, can be dropped...

ferdonline replied:
dropped

pramodk commented on var/spack/repos/builtin/packages/sim-model/package.py:
```diff
@@ -0,0 +1,147 @@
+    def setup_environment(self, spack_env, run_env):
+        spack_env.unset('LC_ALL')
+        run_env.set('LIBNRNMECH_PATH', self.spec.prefix.lib.join('libnrnmech.so'))
+        run_env.set('BGLIBPY_MOD_LIBRARY_PATH', self.spec.prefix.lib.join('libnrnmech.so'))
```
what's different between two? can't we just have one? for example `NRNMECH_LIB_PATH` ?

ferdonline replied:
I wrote in the PR. Temporarily, for compat with bglibpy, we should have BGLIBPY_MOD_LIBRARY_PATH. When they change to the new name we drop it.
I will change LIBNRNMECH_PATH to NRNMECH_LIB_PATH

ferdonline replied:
done

ferdonline committed:
Addressed PK comments and little cleanup

ferdonline committed:
Adding NEURON_INIT_MPI to env to init MPI (python)

pramodk commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -446,6 +448,40 @@ def force_remove(*paths):
+        pth = join_path(src, name)
+        isdir(pth) or copyfunc(pth, dst)
+
+def makelink(src, dst):
```
minor : just wondering if this should be `make_link`

pramodk commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -21,147 +20,88 @@ class NeurodamusModel(Package):
-                prefix.bin.special)
+        # base dest dirs already created by model install
+        arch = arch = spec.architecture.target
+        dst_libname = 'libnrnmechdamus.so'
```
feel weird whenever I see `damus` :) I would suggest `neurodamus`, 5 extra chars.

pramodk reviewed (approved):
looks good to me..we can do some refinements later.

ferdonline committed:
reNamings and doc fix...

ferdonline committed:
Setting env vars to help user compile more mods

ferdonline committed:
FIX: special-core to link only with nrnmech_nd

ferdonline commented:
Full SimulationStack tests passing.
https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/506/pipeline/16

ferdonline committed:
Upcase ND_INCFLAGS ND_LOADFLAGS
